### PR TITLE
✅  Test DataAccessPreferenceKeys

### PR DESCRIPTION
**Pull request by aldebaran30701**
### Description of the Change
Adds test cases for preference keys. 
Found small improvements to make, like actually setting the previous save directory to the previous saved, rather than duplicating the same directory.
Also added a check to make sure that the saved preference was a directory and not a file before saving it to preferences.
Switched usage of "" to StringUtils.EMPTY for consistency.

NOTE: Some tests contain duplicated code as I had trouble mocking two static methods of the same class. To get around this, I mocked only the internals of each static method and this worked perfect.

### CODE REVIEW

File: CoreDataAccessView/src/au/gov/asd/tac/constellation/views/dataaccess/state/DataAccessPreferenceKeys.java [hunk 1]
```diff
@@ -69,19 +70,22 @@ public static File getDataAccessResultsDirEx() {
...
     public static void setDataAccessResultsDir(final File dir) {
         final Preferences prefs = NbPreferences.forModule(DataAccessPreferenceKeys.class);
-        prefs.put(SAVE_DATA_DIR_PREF, dir == null ? "" : dir.getAbsolutePath());
-        if (dir != null) {
-            prefs.put(PREVIOUS_DATA_DIR_PREF, dir.getAbsolutePath());
+        if (dir != null && dir.isDirectory()) {
+            prefs.put(PREVIOUS_DATA_DIR_PREF, prefs.get(SAVE_DATA_DIR_PREF, ""));
         }
+        prefs.put(SAVE_DATA_DIR_PREF, dir == null || !dir.isDirectory() ? "" : dir.getAbsolutePath());
+
```
Comment (aldebaran30701) [inline CoreDataAccessView/src/au/gov/asd/tac/constellation/views/dataaccess/state/DataAccessPreferenceKeys.java:88]: Added a check to ensure that the dir parameter is actually a dir (consistent with other methods)
Ensured that the previous directory preference was actually going to be saved as the previous (before this change it was storing the same value twice.)

File: CoreDataAccessView/src/au/gov/asd/tac/constellation/views/dataaccess/state/DataAccessPreferenceKeys.java [hunk 2]
```diff
@@ -109,11 +113,11 @@ public static File getPreviousDataAccessResultsDir() {
      *
      * @return The preference as a directory; null if the directory is not set.
      */
-    private static File getDir(final String pref) {
+    protected static File getDir(final String pref) {
         final Preferences prefs = NbPreferences.forModule(DataAccessPreferenceKeys.class);
         final String s = prefs.get(pref, "");
 
-        return !s.isEmpty() ? new File(s) : null;
+        return StringUtils.isNotEmpty(s) ? new File(s) : null;
```
Comment (aldebaran30701) [inline CoreDataAccessView/src/au/gov/asd/tac/constellation/views/dataaccess/state/DataAccessPreferenceKeys.java:120]: Null pointer was raised, so safe check for emptiness was used

File: CoreDataAccessView/test/unit/src/au/gov/asd/tac/constellation/views/dataaccess/state/DataAccessPreferenceKeysNGTest.java [hunk 3]
```diff
@@ -0,0 +1,676 @@
...
+    }
+
+    @AfterClass
+    public static void tearDownClass() throws Exception {
+    }
+
+    @BeforeMethod
+    public void setUpMethod() throws Exception {
+        // Before each method, open the mocks.
+        MockitoAnnotations.openMocks(this);
```
Comment (antares1470) [inline CoreDataAccessView/test/unit/src/au/gov/asd/tac/constellation/views/dataaccess/state/DataAccessPreferenceKeysNGTest.java:65]: I think this only needs to be called if you are using Mock annotations which it doesn't appear so. Correct me if I'm wrong

File: CoreDataAccessView/test/unit/src/au/gov/asd/tac/constellation/views/dataaccess/state/DataAccessPreferenceKeysNGTest.java [hunk 4]
```diff
@@ -0,0 +1,676 @@
...
+
+                // Test saveResultsEnabled with invalid DataAccessResultsDirectory
+                boolean expResult2 = false;
+                boolean result2 = DataAccessPreferenceKeys.isSaveResultsEnabled();
+                assertEquals(result2, expResult2);
+            }
+        } catch (final IOException ex) {
+            // Fail when test cannot create a file
+            Exceptions.printStackTrace(ex);
+            fail();
```
Comment (antares1470) [inline CoreDataAccessView/test/unit/src/au/gov/asd/tac/constellation/views/dataaccess/state/DataAccessPreferenceKeysNGTest.java:184]: I would just throw the exception

File: CoreDataAccessView/test/unit/src/au/gov/asd/tac/constellation/views/dataaccess/state/DataAccessPreferenceKeysNGTest.java [hunk 5]
```diff
@@ -0,0 +1,676 @@
...
+                assertEquals(result, tempFile);
+
+                // Verify no file is returned because it needs to be a directory
+                final File expResult = null;
+                assertEquals(DataAccessPreferenceKeys.getDataAccessResultsDir(), expResult);
+            }
+        } catch (final IOException ex) {
+            // Fail when test cannot create a file
+            Exceptions.printStackTrace(ex);
+            fail();
```
Comment (antares1470) [inline CoreDataAccessView/test/unit/src/au/gov/asd/tac/constellation/views/dataaccess/state/DataAccessPreferenceKeysNGTest.java:266]: same here

File: CoreDataAccessView/test/unit/src/au/gov/asd/tac/constellation/views/dataaccess/state/DataAccessPreferenceKeysNGTest.java [hunk 6]
```diff
@@ -0,0 +1,676 @@
...
+                final File expResult = null;
+                assertEquals(DataAccessPreferenceKeys.getDataAccessResultsDir(), expResult);
+
+                // Verify that a file is returned because it does not care about directory.
+                assertEquals(DataAccessPreferenceKeys.getDataAccessResultsDirEx(), tempFile);
+            }
+        } catch (final IOException ex) {
+            // Fail when test cannot create a file
+            Exceptions.printStackTrace(ex);
+            fail();
```
Comment (antares1470) [inline CoreDataAccessView/test/unit/src/au/gov/asd/tac/constellation/views/dataaccess/state/DataAccessPreferenceKeysNGTest.java:402]: ditto

File: CoreDataAccessView/test/unit/src/au/gov/asd/tac/constellation/views/dataaccess/state/DataAccessPreferenceKeysNGTest.java [hunk 7]
```diff
@@ -0,0 +1,676 @@
...
+            dir = File.createTempFile("testfile", ".txt");
+
+            DataAccessPreferenceKeys.setDataAccessResultsDir(dir);
+
+            assertEquals(preferences.get(currentDir, null), "");
+            assertEquals(preferences.get(previousDir, null), null);
+
+        } catch (final IOException ex) {
+            Exceptions.printStackTrace(ex);
+            fail();
```
Comment (antares1470) [inline CoreDataAccessView/test/unit/src/au/gov/asd/tac/constellation/views/dataaccess/state/DataAccessPreferenceKeysNGTest.java:467]: ditto

File: CoreDataAccessView/test/unit/src/au/gov/asd/tac/constellation/views/dataaccess/state/DataAccessPreferenceKeysNGTest.java [hunk 8]
```diff
@@ -0,0 +1,676 @@
...
+                assertEquals(result, tempFile);
+
+                // Verify no file is returned because it needs to be a directory
+                final File expResult = null;
+                assertEquals(DataAccessPreferenceKeys.getPreviousDataAccessResultsDir(), expResult);
+            }
+        } catch (final IOException ex) {
+            // Fail when test cannot create a file
+            Exceptions.printStackTrace(ex);
+            fail();
```
Comment (antares1470) [inline CoreDataAccessView/test/unit/src/au/gov/asd/tac/constellation/views/dataaccess/state/DataAccessPreferenceKeysNGTest.java:544]: ditto

**Review by antares1470 — CHANGES_REQUESTED** [review]

File: CoreDataAccessView/src/au/gov/asd/tac/constellation/views/dataaccess/state/DataAccessPreferenceKeys.java (hunk 2 above)
Comment (arcturus2) [reply CoreDataAccessView/src/au/gov/asd/tac/constellation/views/dataaccess/state/DataAccessPreferenceKeys.java:120]: great, so the unit test found some issues and you fixed them as well. good on you @aldebaran30701

**Review by antares1470 — APPROVED** [review]

File: CoreDataAccessView/test/unit/src/au/gov/asd/tac/constellation/views/dataaccess/state/DataAccessPreferenceKeysNGTest.java [hunk 9]
```diff
@@ -0,0 +1,649 @@
...
+                // Verify no file is returned because it needs to be a directory
+                final File expResult = null;
+                assertEquals(DataAccessPreferenceKeys.getPreviousDataAccessResultsDir(), expResult);
+            }
+        } finally {
+            // Cleanup
+            if (tempFile != null && tempFile.exists()) {
+                tempFile.delete();
+            }
+        }
```
Comment (arcturus2) [inline CoreDataAccessView/test/unit/src/au/gov/asd/tac/constellation/views/dataaccess/state/DataAccessPreferenceKeysNGTest.java:523]: really like this!

**Review by arcturus2 — APPROVED** [review]

**Pull request merged**